### PR TITLE
Add timestamp query in AuthenticationClient::SignInClient 

### PR DESCRIPTION
**Pull request by meravingen123**
* Add additional timestamp https request while using
   SignInClient method
 * Change token expiriation checking to use steady clock
   instead of system clock.
 * Add new public API to TokenResult and SignInResult to
   get expiriation time of token in seconds.

Resolves: OLPEDGE-837

Signed-off-by: Serhii Lozynskyi <ext-serhii.lozynskyi@here.com>

### CODE REVIEW

File: olp-cpp-sdk-authentication/include/olp/authentication/SignInResult.h [hunk 1]
```diff
@@ -107,7 +109,15 @@ class AUTHENTICATION_API SignInResult {
    * @return The Epoch time when the token expires, or -1 if the token is
    * invalid.
    */
+  OLP_SDK_DEPRECATED("Use GetExpiryTimeInSeconds")
   time_t GetExpiryTime() const;
+  /**
+   * @brief Gets the access token expiry time in seconds.
+   *
+   * @return The Epoch time when the token expires, or -1 if the token is
```
Comment (r0busta) [inline olp-cpp-sdk-authentication/include/olp/authentication/SignInResult.h]: Is it Epoch time or sort of duration in seconds?

File: olp-cpp-sdk-authentication/src/AuthenticationClient.cpp [hunk 2]
```diff
@@ -179,10 +181,14 @@ class AuthenticationClient::Impl final {
   void SetTaskScheduler(std::shared_ptr<TaskScheduler> task_scheduler);
 
  private:
+  using TimeResponse = client::ApiResponse<time_t, client::ApiError>;
+  using TimeCallback = std::function<void(TimeResponse)>;
```
Comment (r0busta) [inline olp-cpp-sdk-authentication/src/AuthenticationClient.cpp]: Would it make sense to move them closer to the usage in the SignInClient?

File: olp-cpp-sdk-authentication/src/AuthenticationClient.cpp [hunk 3]
```diff
@@ -264,97 +270,182 @@ client::CancellationToken AuthenticationClient::Impl::SignInClient(
...
-          }
-          default:
-            callback(response);
-            break;
+        auto timestamp_it = document->FindMember("timestamp");
+        if (timestamp_it == document->MemberEnd() ||
+            !timestamp_it->value.IsUint()) {
+          callback(client::ApiError(
+              client::ErrorCode::InternalFailure,
+              "JSon document must contain timestamp integer field"));
```
Comment (r0busta) [inline olp-cpp-sdk-authentication/src/AuthenticationClient.cpp]:
```suggestion
              "JSON document must contain timestamp integer field"));
```

File: olp-cpp-sdk-authentication/src/AuthenticationClient.cpp [hunk 4]
```diff
@@ -264,97 +270,182 @@ client::CancellationToken AuthenticationClient::Impl::SignInClient(
...
-            cache->locked([credentials, &response](
-                              utils::LruCache<std::string, SignInResult>& c) {
-              return c.InsertOrAssign(credentials.GetKey(), response);
-            });
-            // intentially do not break
-          }
-          default:
-            callback(response);
-            break;
+        auto timestamp_it = document->FindMember("timestamp");
```
Comment (r0busta) [inline olp-cpp-sdk-authentication/src/AuthenticationClient.cpp]:
```suggestion
        const auto timestamp_it = document->FindMember("timestamp");
```

File: olp-cpp-sdk-authentication/src/SignInResultImpl.h [hunk 5]
```diff
@@ -96,6 +100,8 @@ class SignInResultImpl : public BaseResult {
   std::string token_type_;
   std::string client_id_;
   time_t expiry_time_;
+  std::chrono::seconds expiry_time_seconds_;
```
Comment (andescu) [inline olp-cpp-sdk-authentication/src/SignInResultImpl.h]:
```suggestion
  std::chrono::seconds expires_in_;
```

File: olp-cpp-sdk-authentication/src/AuthenticationClient.cpp [hunk 6]
```diff
@@ -179,10 +181,14 @@ class AuthenticationClient::Impl final {
   void SetTaskScheduler(std::shared_ptr<TaskScheduler> task_scheduler);
 
  private:
+  using TimeResponse = client::ApiResponse<time_t, client::ApiError>;
+  using TimeCallback = std::function<void(TimeResponse)>;
+  client::CancellationToken GetTimeFromServer(TimeCallback callback);
```
[the remark this answers is not in the document]
Comment (meravingen123) [reply olp-cpp-sdk-authentication/src/AuthenticationClient.cpp]: Yes as this is called from AuthenticationClient::Impl::SignInClient() which is async also.

File: olp-cpp-sdk-authentication/src/AuthenticationClient.cpp (hunk 2 above)
Comment (meravingen123) [reply olp-cpp-sdk-authentication/src/AuthenticationClient.cpp]: What do you mean by that? they are near GetTimeFromServer() method and are used in declaration of that method.

File: olp-cpp-sdk-authentication/src/AutoRefreshingToken.cpp [hunk 7]
```diff
@@ -30,10 +30,10 @@
...
 
-std::chrono::system_clock::time_point ComputeRefreshTime(
-    time_t expiry_time, const std::chrono::seconds& minimum_validity) {
-  auto expiry_time_chrono = std::chrono::system_clock::from_time_t(expiry_time);
-  auto now = std::chrono::system_clock::now();
+std::chrono::steady_clock::time_point ComputeRefreshTime(
+    std::chrono::seconds expiry_time_seconds, const std::chrono::seconds& minimum_validity) {
+  auto now = std::chrono::steady_clock::now();
+  auto expiry_time_chrono = now + expiry_time_seconds;
   return (expiry_time_chrono <= now) ? now
```
[the remark this answers is not in the document]
Comment (meravingen123) [reply olp-cpp-sdk-authentication/src/AutoRefreshingToken.cpp]: minimum_validity time is used to refresh token before it expires. There is no need to pass big numbers there like 24h. Naming is kinda bad for it. probably we should rename it in my opinion.

File: olp-cpp-sdk-authentication/src/AuthenticationClient.cpp [hunk 8]
```diff
@@ -264,97 +270,182 @@ client::CancellationToken AuthenticationClient::Impl::SignInClient(
...
-          default:
-            callback(response);
-            break;
+        auto timestamp_it = document->FindMember("timestamp");
+        if (timestamp_it == document->MemberEnd() ||
+            !timestamp_it->value.IsUint()) {
+          callback(client::ApiError(
+              client::ErrorCode::InternalFailure,
+              "JSon document must contain timestamp integer field"));
+          return;
```
[the remark this answers is not in the document]
Comment (meravingen123) [reply olp-cpp-sdk-authentication/src/AuthenticationClient.cpp]: what do you mean by that? is that really necessary here?

File: olp-cpp-sdk-authentication/src/AuthenticationClient.cpp (hunk 2 above)
Comment (r0busta) [reply olp-cpp-sdk-authentication/src/AuthenticationClient.cpp]: Pardon me. Didn't notice that `TimeResponse` in is the declaration down there.

**Comment by r0busta** [comment]
Is still a POC? If not, please consider creating a proper commit message with all the bells&wistles :)

**Comment by r0busta** [comment]
Thanks for editing the commit message 👍

One minor point though. Please prefer using imperative-mood in commit messages. It's probably the most popular style in the open-source world, and it also makes sense from the perspective of readability of git repository history and logs.

Here's what `git` authors [have to say on this](https://git.kernel.org/pub/scm/git/git.git/tree/Documentation/SubmittingPatches?id=HEAD#n133):

> Describe your changes in imperative mood, e.g. "make xyzzy do frotz" instead of "[This patch] makes xyzzy do frotz" or "[I] changed xyzzy to do frotz", as if you are giving orders to the codebase to change its behavior.

**Pull request merged**